Include dependency role names in `role_names`.

Xaroth opened the pull request:
-Change: Include dependency role names in `role_names`.
-Add: `play_role_names` magic variable to include only explicitly named roles (formerly `role_names`).
-Add: `dependent_role_names` magic variable to include all dependency names for all roles.

##### SUMMARY
As per #16068, the `role_names` magic variable only contains the role names of roles explicitly listed in the current play. This PR is to change this functionality to be aware of dependencies.

The main rationale is that once your tree of roles grows, you inevitably introduce conflicts between roles, to prevent these conflicts from happening, one would want to add an early failure when a role is being installed that conflicts with an already existing one.
The easiest way to do this is to check what other roles are active, and notify accordingly. This would work with `role_names`, but **only** if the role was explicitly listed in the playbook. If the role was included by means of a dependency, it would not show up, and you would have to do other manual checking to prevent this.

Therefore we introduce 2 new magic variables, and redefine one:

- `role_names` now includes all role names, including those of dependencies. The most common use case when checking in role_names is to see if the target host is affected by a specific role. The addition of dependencies to this list would not break those use cases, but instead would make them more accurate.
- `play_role_names` assumes the role of the former `role_names`, only listing explicitly listed roles as per the play. Should there be a requirement to see if a role is explicitly configured, this list can be used
- `dependent_role_names` lists all roles that are included because they are a dependency of another role. This means they may include roles that are also defined in `play_role_names`, so a check against `play_role_names` can verify whether a specific role is _only_ a dependency, or _both_ a dependency _and_ explicitly required.

Documentation has been adjusted (and added) for these magic variables.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
Magic Variables

##### ANSIBLE VERSION
```paste below
ansible 2.7.0rc4.post0
```
Please note that this affects ansible versions 2.0.0 up until `master`

##### ADDITIONAL INFORMATION
Given a setup of 2 roles (A and B, for ease of use), where A/meta/main.yml lists B as a dependency, `role_names` would show `['A']`. This makes it impossible to detect whether or not B will be installed on a host.

After this change, `role_names` would show `['A', 'B']`, `play_role_names` shows `['A']`, and `dependent_role_names` shows `['B']`

Xaroth committed:
-Change: Include dependency role names in `role_names`.
-Add: `play_role_names` magic variable to include only explicitly named roles (formerly `role_names`).
-Add: `dependent_role_names` magic variable to include all dependency names for all roles.

bcoca commented:
FYI, we require all new 'magic vars' to be `ansible_` prefixed

Xaroth commented:
That's doable; but that leaves the case of the existing role_names item, should that be renamed to ansible_role_names or left as-is?

bcoca commented:
keep as is for backwards compat, create new prefixed one that does what your update does

Xaroth committed:
-Change: use the ansible_ prefix for new magic variables.
-Change: keep `role_names` as a deprecated variable, using the old functionality.

Xaroth commented:
@bcoca I changed the functionality as you requested:

`ansible_*` with the new functionality
`role_names` with the old functionality

Documentation updated to mark `role_names` as deprecated in favor of `ansible_role_names` or `ansible_play_role_names`.

I did not see a clear way to add deprecation warnings to the usage of `role_names`, however, I saw that `play_hosts` does not use this either, so I left it similar to that.

Xaroth commented:
Not sure why the checks have failed, looks more like an error on the CI side than on the code side.

Commit:
-Add: changelog fragment for the role_names rework.

gundalow reviewed (commented):
Please add tests :)

Xaroth committed:
-Add: Tests for the role_names (and ansible_*role_names) special variables

Xaroth commented:
Thanks @gundalow , tests have been added.

bcoca commented on docs/docsite/rst/reference_appendices/special_variables.rst:
```diff
@@ -79,14 +90,18 @@ play_hosts
 ansible_play_name
     The name of the currently executed play. Added in ``2.8``.
 
+play_role_names
```
don't create a new variable that is not conforming to the `ansilbe_` pattern

Xaroth replied:
Hm, this seems to be an error introduced by rebasing, will correct.

Xaroth replied:
This should have been corrected with the last commit.

bcoca commented on lib/ansible/vars/manager.py:
```diff
@@ -451,7 +451,23 @@ def _get_magic_variables(self, play, host, task, include_hostvars, include_deleg
+            variables['ansible_dependent_role_names'] = dependency_role_names
+
+            # DEPRECATED: role_names should be deprecated in favor of ansible_role_names or ansible_play_role_names
+            variables['role_names'] = variables['ansible_play_role_names']
```
as in previous comment, don't change `role_names` the new variable provides what you want, don't change the existing variable as people are using the current behavior.

Xaroth replied:
The functionality of role_names remains the same before- and after-change; I merely moved the primary point of assignment to `ansible_play_role_names` and made `role_names` a copy from that.

This behavior is a copy from lines 491-493: https://github.com/ansible/ansible/blob/884ff74353a327165b9f4afcdae5081d006173ec/lib/ansible/vars/manager.py#L491-L493

bcoca replied:
ah, misread, thought you were making it same as ansible_role_names

Xaroth committed:
-Fix: resolve erroneous documentation snippet that was introduced after rebasing.

acozine commented on changelogs/fragments/46483-role_names-change.yaml:
```diff
@@ -0,0 +1,11 @@
+minor_changes:
+  - magic variables - added a new ``ansible_role_names`` magic variable to include the names of roles being applied to
+    the host both directly and indirectly (via dependencies).
+  - magic variabels - added a new ``ansible_play_role_names`` magic variable to mimic the old functionality of
```
typo - `variabels` - though as a new word, it sounds pretty . . .

Xaroth replied:
Thanks, fixed.

acozine commented:
Docs look good, thanks @Xaroth.

Xaroth committed:
-Fix: explicitly sort to ensure list comparison works in test.
